File: Reqnroll/Chapter14/14.4-TerminologyPromotionPattern/After/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.App.Services;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(OrderingContext orderingContext, OrderService orderService)
{
    [When("the customer places an order for {int} pizza(s) of size {pizza-size}")]
    public void WhenTheCustomerPlacesAnOrderFor(int count, PizzaSize pizzaSize)
    {
        var order = new OrderObjectMother()
            .WithItems(count, size: pizzaSize)
            .Build();

        AuthenticationService.Login(DomainDefaults.CustomerName);
        var placedOrder = orderService.PlaceOrder(order);
        orderingContext.CurrentOrderNo = placedOrder.OrderNo;
    }

    #region Additional Step Definitions

    [Then("the order should be rejected")]
    public void ThenTheOrderShouldBeRejected()
    {
        int orderNo = orderingContext.CurrentOrderNo ?? throw new InvalidOperationException("No current order");
        var order = orderService.GetOrder(orderNo) ?? throw new InvalidOperationException("Order not found");
        Assert.AreEqual(OrderStatus.Rejected, order.Status);
    }

    #endregion
}

File: CucumberJVM/Chapter14/14.4-TerminologyPromotionPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.OrderStatus;
import com.wimp.app.models.PizzaSize;
import com.wimp.app.services.AuthenticationService;
import com.wimp.app.services.OrderService;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.OrderObjectMother;
import com.wimp.app.specs.support.OrderingContext;
import io.cucumber.java.en.*;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

public class OrderingStepDefinitions {
    private final OrderingContext orderingContext;
    private final OrderService orderService;

    public OrderingStepDefinitions(OrderingContext orderingContext, OrderService orderService) {
        this.orderingContext = orderingContext;
        this.orderService = orderService;
    }

    @When("the customer places an order for {int} pizza(s) of size {string}")
    public void theCustomerPlacesAnOrderForPizzasOfSize(int count, String pizzaSize) {
        var order = new OrderObjectMother().withItems(count, PizzaSize.valueOf(pizzaSize.toUpperCase())).build();

        AuthenticationService.login(DomainDefaults.CUSTOMER_NAME);
        var placedOrder = orderService.placeOrder(order);
        orderingContext.setCurrentOrderNo(placedOrder.getOrderNo());
    }

    @Then("the order should be rejected")
    public void theOrderShouldBeRejected() {
        int orderNo = orderingContext.getCurrentOrderNoVerified();
        var order = Optional.of(orderService.getOrder(orderNo))
            .orElseThrow(() -> new RuntimeException("Order not found"));
        assertEquals(OrderStatus.REJECTED, order.getStatus());
    }
}
